src/main/java/com/zendesk/maxwell/schema/SchemaStore.java

```diff
@@ -0,0 +1,197 @@
+		InputStream schemaSQL = getClass().getResourceAsStream(
+				"/sql/maxwell_schema.sql");
+		BufferedReader r = new BufferedReader(new InputStreamReader(schemaSQL));
+		String sql = "", line;
+
+		LOGGER.info("Creating maxwell database");
+		while ((line = r.readLine()) != null) {
+			sql += line + "\n";
+		}
+
+		for (String statement : StringUtils.splitByWholeSeparator(sql, "\n\n")) {
+			if (statement.length() == 0)
+				continue;
+
+			this.connection.createStatement().execute(statement);
+		}
+	}
+
+	public static SchemaStore restore(Connection connection,
+			BinlogPosition targetPosition) throws SQLException, SchemaSyncError {
+		SchemaStore s = new SchemaStore(connection);
+
+		s.restoreFrom(targetPosition);
+
+		return s;
+	}
+
+	private void restoreFrom(BinlogPosition targetPosition)
+			throws SQLException, SchemaSyncError {
+		PreparedStatement p;
+		ResultSet schemaRS = findSchemaID(targetPosition);
+
+		if (schemaRS == null)
+			throw new SchemaSyncError("Could not find schema for "
+					+ targetPosition.getFile() + ":"
+					+ targetPosition.getOffset());
+
+		ArrayList<Database> databases = new ArrayList<>();
+		this.schema = new Schema(databases);
+		this.position = new BinlogPosition(schemaRS.getInt("binlog_position"),
+				schemaRS.getString("binlog_file"));
+
+		p = connection.prepareStatement("SELECT * from `maxwell`.`databases` where schema_id = ? ORDER by id");
+		p.setInt(1, schemaRS.getInt("id"));
+		ResultSet dbRS = p.executeQuery();
+
+		while (dbRS.next()) {
+			this.schema.getDatabases().add(restoreDatabase(dbRS.getInt("id"), dbRS.getString("name"), dbRS.getString("encoding")));
+		}
+	}
+
+	private Database restoreDatabase(int id, String name, String encoding) throws SQLException {
+		Statement s = connection.createStatement();
+		Database d = new Database(name, encoding);
+
+		ResultSet tRS = s.executeQuery("SELECT * from `maxwell`.`tables` where database_id = " + id + " ORDER by id");
+
+		while (tRS.next()) {
+			String tName = tRS.getString("name");
+			String tEncoding = tRS.getString("encoding");
+			int tID = tRS.getInt("id");
+
+			restoreTable(d, tName, tID, tEncoding);
+		}
+		return d;
+	}
+
+	private void restoreTable(Database d, String name, int id, String encoding) throws SQLException {
+		Statement s = connection.createStatement();
+		Table t = d.buildTable(name, encoding);
+
+		ResultSet cRS = s.executeQuery("SELECT * from `maxwell`.`columns` where table_id = " + id + " ORDER by id");
+
+		int i = 0;
+		while (cRS.next()) {
+			ColumnDef c = ColumnDef.build(t.getName(),
+					cRS.getString("name"), cRS.getString("encoding"),
+					cRS.getString("coltype"), i++,
+					cRS.getInt("is_signed") == 1);
+			t.getColumnList().add(c);
+		}
+	}
+
+	private ResultSet findSchemaID(BinlogPosition targetPosition)
+			throws SQLException {
+		PreparedStatement s = connection
+				.prepareStatement("SELECT * from `maxwell`.`schemas` where binlog_file <= ? and binlog_position <= ? "
```

Comment:
when you say "offset", do you mean the binlog position?  it doesn't, it refers to a very literal file offset on disk